cmd/osm-controller/osm-controller.go

```diff
@@ -210,17 +210,35 @@ func main() {
 	// TODO(draychev): figure out the NS and POD
 	metricsStore := metricsstore.NewMetricStore("TBD_NameSpace", "TBD_PodName")
 
-	// Expose /debug endpoints and data only if the enableDebugServer flag is enabled
-	var debugServer debugger.DebugServer
-	if cfg.IsDebugServerEnabled() {
-		debugServer = debugger.NewDebugServer(certDebugger, xdsServer, meshCatalog, kubeConfig, kubeClient, cfg, kubernetesClient)
-	}
-
 	funcProbes := []health.Probes{xdsServer}
 	httpProbes := getHTTPHealthProbes()
-	httpServer := httpserver.NewHTTPServer(funcProbes, httpProbes, metricsStore, constants.MetricsServerPort, debugServer)
+	httpServer := httpserver.NewHTTPServer(funcProbes, httpProbes, metricsStore, constants.MetricsServerPort)
 	httpServer.Start()
 
+	// Expose /debug endpoints and data only if the enableDebugServer flag is enabled
+	debugServerRunning := false
+	debugImpl := debugger.NewDebugImpl(certDebugger, xdsServer, meshCatalog, kubeConfig, kubeClient, cfg, kubernetesClient)
+	debugServer := httpserver.NewDebugServer(debugImpl, constants.DebugPort)
+
+	if cfg.IsDebugServerEnabled() {
+		debugServerRunning = true
+		debugServer.Start()
+	}
+	go func() {
+		for change := range cfg.GetAnnouncementsChannel() {
```

Comment:
This seems a bit hacky to me given that there is no way to know a config change is meant for the debugger. This is not going to be a problem now because no other component is listening to announcements from `Configurator`, but I would look into implementing a per event notification channel to solve such issues.

At the very least, please drop a comment here regarding why this even works as expected.

Reply:
You're right it'll run for every change to ConfigMap, for now it works but in the future like you mentioned we could implement individual notification channels.
It works because we check to see what the value for enableDebugServer is with `cfg.IsDebugServerEnabled()`.
We also check debugServerRunning so that if a different field in ConfigMap is changed it won't do anything to the debug server like if `debugServerRunning=false` and  in ConfigMap a field like `egress` is changed and `enable_debug_server = false` (unchanged) then we won't try to run `debugServer.Stop()`